source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2232 @@
+    struct aws_allocator *allocator = task_args->allocator;
+    struct nw_socket *nw_socket = task_args->nw_socket;
+
+    if (status != AWS_TASK_STATUS_CANCELED) {
+        s_lock_base_socket(nw_socket);
+        struct aws_socket *socket = nw_socket->base_socket_synced_data.base_socket;
+        if (task_args->written_fn) {
+            task_args->written_fn(socket, task_args->error_code, task_args->bytes_written, task_args->user_data);
+        }
+        s_unlock_base_socket(nw_socket);
+    }
+
+    s_socket_release_internal_ref(nw_socket);
+
+    aws_mem_release(allocator, task_args);
+}
+
+static void s_handle_write_fn(
+    struct nw_socket *nw_socket,
+    int error_code,
+    size_t bytes_written,
+    void *user_data,
+    aws_socket_on_write_completed_fn *written_fn) {
+    AWS_FATAL_ASSERT(s_validate_event_loop(nw_socket->event_loop));
+
+    struct nw_socket_written_args *args =
+        aws_mem_calloc(nw_socket->allocator, 1, sizeof(struct nw_socket_written_args));
+
+    args->nw_socket = nw_socket;
+    args->allocator = nw_socket->allocator;
+    args->error_code = error_code;
+    args->written_fn = written_fn;
+    args->user_data = user_data;
+    args->bytes_written = bytes_written;
+    s_socket_acquire_internal_ref(nw_socket);
+
+    aws_task_init(&args->task, s_process_write_task, args, "writtenTask");
+
+    aws_event_loop_schedule_task_now(nw_socket->event_loop, &args->task);
+}
+
+static int s_socket_connect_fn(
+    struct aws_socket *socket,
+    const struct aws_socket_endpoint *remote_endpoint,
+    struct aws_event_loop *event_loop,
+    aws_socket_on_connection_result_fn *on_connection_result,
+    void *user_data) {
+    struct nw_socket *nw_socket = socket->impl;
+
+    AWS_FATAL_ASSERT(event_loop);
+    AWS_FATAL_ASSERT(!socket->event_loop);
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_IO_SOCKET, "id=%p handle=%p: beginning connect.", (void *)socket, socket->io_handle.data.handle);
+
+    // Apple Network Framework uses a connection based abstraction on top of the UDP layer. We should always do an
+    // "connect" action after aws_socket_init() regardless it's a UDP socket or a TCP socket.
+    AWS_FATAL_ASSERT(on_connection_result);
+    s_lock_socket_synced_data(nw_socket);
+    if (nw_socket->synced_data.state != INIT) {
+        s_unlock_socket_synced_data(nw_socket);
+        return aws_raise_error(AWS_IO_SOCKET_ILLEGAL_OPERATION_FOR_STATE);
+    }
+
+    /* fill in posix sock addr, and then let Network framework sort it out. */
+    size_t address_strlen;
+    if (aws_secure_strlen(remote_endpoint->address, AWS_ADDRESS_MAX_LEN, &address_strlen)) {
+        s_unlock_socket_synced_data(nw_socket);
+        AWS_LOGF_DEBUG(
+            AWS_LS_IO_SOCKET,
+            "id=%p handle=%p: failed to parse address %s:%d.",
+            (void *)socket,
+            socket->io_handle.data.handle,
+            remote_endpoint->address,
+            (int)remote_endpoint->port);
+        return AWS_OP_ERR;
```

Comment:
We should probably raise an error here in case anything is checking for last error on a failed operation.